lib/container/emqx_container.ex

```diff
@@ -0,0 +1,150 @@
+      iex> new_config = EmqxContainer.with_image(config, "emqx:xyz")
+      iex> new_config.image
+      "emqx:xyz"
+  """
+  def with_image(%__MODULE__{} = config, image) when is_binary(image) do
+    %{config | image: image}
+  end
+
+  def with_ports(
+        %__MODULE__{} = config,
+        mqtt_port \\ @default_mqtt_port,
+        mqtts_port \\ @default_mqtts_port,
+        mqtt_over_ws_port \\ @default_mqtt_over_ws_port,
+        mqtt_over_wss_port \\ @default_mqtt_over_wss_port,
+        dashboard_port \\ @default_dashboard_port
+      ) do
+    %{
+      config
+      | mqtt_port: mqtt_port,
+        mqtts_port: mqtts_port,
+        mqtt_over_ws_port: mqtt_over_ws_port,
+        mqtt_over_wss_port: mqtt_over_wss_port,
+        dashboard_port: dashboard_port
+    }
+  end
+
+  @doc """
+  Retrieves the default Docker image for the Emqx container.
+  """
+  def default_image, do: @default_image
+
+  @doc """
+  Returns the address on the _host machine_ where the Emqx container is listening.
+  """
+  def host, do: Testcontainers.get_host()
+
+  @doc """
+  Returns the port on the _host machine_ where the Emqx container is listening.
+  """
+  def mqtt_port(%Container{} = container),
+    do: Container.mapped_port(container, @default_mqtt_port)
+
+  defimpl ContainerBuilder do
+    import Container
+
+    @doc """
+    Builds a new container instance based on the provided configuration.
+    """
+    @impl true
+    def build(%EmqxContainer{} = config) do
+      if not String.starts_with?(config.image, EmqxContainer.default_image()) do
+        raise ArgumentError,
+          message: "Image #{config.image} is not compatible with #{EmqxContainer.default_image()}"
+      end
+
+      new(config.image)
+      |> with_exposed_ports(exposed_ports(config))
+      |> with_waiting_strategies(waiting_strategies(config))
+    end
+
+    defp exposed_ports(config),
+      do: [
+        config.mqtt_port,
+        config.mqtts_port,
+        config.mqtt_over_ws_port,
+        config.mqtt_over_wss_port,
+        config.dashboard_port
+      ]
+
+    defp waiting_strategies(config),
+      do: [
+        PortWaitStrategy.new(EmqxContainer.host(), config.mqtt_port, config.wait_timeout, 1000),
+        PortWaitStrategy.new(EmqxContainer.host(), config.mqtts_port, config.wait_timeout, 1000),
+        PortWaitStrategy.new(
+          EmqxContainer.host(),
+          config.mqtt_over_ws_port,
+          config.wait_timeout,
+          1000
+        ),
+        PortWaitStrategy.new(
+          EmqxContainer.host(),
+          config.mqtt_over_wss_port,
+          config.wait_timeout,
+          1000
+        ),
+        PortWaitStrategy.new(
+          EmqxContainer.host(),
+          config.dashboard_port,
+          config.wait_timeout,
+          1000
+        )
+      ]
+
+    @impl true
+    # TODO Implement the `is_starting/3` function for the `ContainerBuilder` protocol.
+    def is_starting(_config, _container, _conn), do: :ok
```

Comment:
i think the wording of this function is a bit wrong. Its actually an "on startup hook" function. Its called when the container "is starting". Used for executing scripts right after container has started. see https://github.com/testcontainers/testcontainers-elixir/blob/main/lib/container/kafka_container.ex#L177

Reply:
Just a FYI, i should think about the naming of the function. If you have a better name feel free to suggest

Reply:
However, i see now the docs in https://github.com/testcontainers/testcontainers-elixir/blob/main/lib/container/kafka_container.ex#L177 is wrong. Its actually not a function that "checks if container is starting" as the kafka implementation says it is.

Reply:
added new issue in https://github.com/testcontainers/testcontainers-elixir/issues/84

Reply:
@jarlah Thanks for your review.

Looking at the startup procedure of the containers, it seems that having the function `is_starting` is redundant, see https://github.com/testcontainers/testcontainers-elixir/blob/b0edf48e5d1a31f13eb312904f5f9c6a94536106/lib/testcontainers.ex#L218.

There are two real implementations, in the aforementioned Kafka container and in the supporting Nginx container (https://github.com/testcontainers/testcontainers-elixir/blob/b0edf48e5d1a31f13eb312904f5f9c6a94536106/test/support/nginx_container.ex#L17).

My suggestion would be to remove this function from the protocol and move the existing implementations into a after-startup hook.

Reply:
we can consider it. Leaving it out of this PR naturally.